File: java/leetcode/src/leetcode/editor/cn/P1026_MaximumDifferenceBetweenNodeAndAncestor.java
//给定二叉树的根节点 root，找出存在于 不同 节点 A 和 B 之间的最大值 V，其中 V = |A.val - B.val|，且 A 是 B 的祖先。 
//
//
// （如果 A 的任何子节点之一为 B，或者 A 的任何子节点是 B 的祖先，那么我们认为 A 是 B 的祖先） 
//
// 
//
// 示例 1： 
//
// 
//
// 
//输入：root = [8,3,10,1,6,null,14,null,null,4,7,13]
//输出：7
//解释： 
//我们有大量的节点与其祖先的差值，其中一些如下：
//|8 - 3| = 5
//|3 - 7| = 4
//|8 - 1| = 7
//|10 - 13| = 3
//在所有可能的差值中，最大值 7 由 |8 - 1| = 7 得出。
// 
//
// 示例 2： 
// 
// 
//输入：root = [1,null,2,null,0,3]
//输出：3
// 
//
// 
//
// 提示： 
//
// 
// 树中的节点数在 2 到 5000 之间。 
// 0 <= Node.val <= 10⁵ 
// 
//
// Related Topics 树 深度优先搜索 二叉树 👍 151 👎 0

package leetcode.editor.cn;

import leetcode.editor.cn.DataStruct.TreeNode;

//Java:节点与其祖先之间的最大差值
//Time:2023-04-18 09:19:13
class P1026_MaximumDifferenceBetweenNodeAndAncestor {
    public static void main(String[] args) {
        Solution solution = new P1026_MaximumDifferenceBetweenNodeAndAncestor().new Solution();
    }
    //leetcode submit region begin(Prohibit modification and deletion)

    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     * int val;
     * TreeNode left;
     * TreeNode right;
     * TreeNode() {}
     * TreeNode(int val) { this.val = val; }
     * TreeNode(int val, TreeNode left, TreeNode right) {
     * this.val = val;
     * this.left = left;
     * this.right = right;
     * }
     * }
     */
    class Solution {
        public int maxAncestorDiff(TreeNode root) {
            if (root == null) {
                return 0;
            }
            return postOrder(root, Integer.MAX_VALUE, Integer.MIN_VALUE);
        }

        public int postOrder(TreeNode node, int min, int max) {
            if (node == null) {
                return max - min;
            }
            int val = node.val;
            int m = postOrder(node.left, Math.min(min, val), Math.max(max, val));
            int n = postOrder(node.right, Math.min(min, val), Math.max(max, val));
            return Math.max(m, n);

        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: python/leetcode/editor/cn/P1026_MaximumDifferenceBetweenNodeAndAncestor.py
from math import inf
from typing import Optional

from leetcode.editor.cn.dataStruct.TreeNode import TreeNode


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        res = 0

        def recursion(node: TreeNode) -> (int, int):
            if not node:
                return inf, -inf
            if not node.left and not node.right:
                return node.val, node.val
            _min1, _max1 = recursion(node.left)
            _min2, _max2 = recursion(node.right)
            _min = min(_min1, _min2)
            _max = max(_max1, _max2)
            nonlocal res
            res = max(res, node.val - _min, _max - node.val)
            return min(_min, node.val), max(_max, node.val)

        recursion(root)
        return res

    def maxAncestorDiff2(self, root: Optional[TreeNode]) -> int:
        res = 0

        def recursion(node: TreeNode, _min, _max) -> (int, int):
            if not node:
                nonlocal res
                res = max(res, _max - _min)
                return
            _min = min(_min, node.val)
            _max = max(_max, node.val)
            recursion(node.left, _min, _max)
            recursion(node.right, _min, _max)

        recursion(root, root.val, root.val)
        return res

# leetcode submit region end(Prohibit modification and deletion)
